<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>Module Hierarchy</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">

  <!-- Tree link -->
      <th bgcolor="#70b0f0" class="navbar-select"
          >&nbsp;&nbsp;&nbsp;Trees&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

      <th class="navbar" width="100%"></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">&nbsp;</td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="module-tree.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<center><b>
 [ <a href="module-tree.html">Module Hierarchy</a>
 | <a href="class-tree.html">Class Hierarchy</a> ]
</b></center><br />
<h1 class="epydoc">Module Hierarchy</h1>
<ul class="nomargin-top">
    <li> <strong class="uidlink"><a href="addSample-module.html">addSample</a></strong>: <em class="summary">This Programme find corresponding MGG and BR32 genes</em>    </li>
    <li> <strong class="uidlink"><a href="buildJobArray-module.html">buildJobArray</a></strong>: <em class="summary">This Programme run bash script in buildJobArray</em>    </li>
    <li> <strong class="uidlink"><a href="buildSNPtoFasta-module.html">buildSNPtoFasta</a></strong>: <em class="summary">This Programme parse GFF, TAB to add SNP in sequences</em>    </li>
    <li> <strong class="uidlink"><a href="buildTracks-module.html">buildTracks</a></strong>: <em class="summary">This Programme map fasta with bwa mem on reference file</em>    </li>
    <li> <strong class="uidlink"><a href="compress-module.html">compress</a></strong>: <em class="summary">This Programme compress all file with pass extention in directory with job array</em>    </li>
    <li> <strong class="uidlink"><a href="concatFastasFile-module.html">concatFastasFile</a></strong>: <em class="summary">This Programme concat multiple fasta files with same sequences name's into uniq file</em>    </li>
    <li> <strong class="uidlink"><a href="countCDSStartStop-module.html">countCDSStartStop</a></strong>: <em class="summary">This Programme count CDS with Start AND/OR Stop or NONE Both, in fasta files</em>    </li>
    <li> <strong class="uidlink"><a href="extractColfromList-module.html">extractColfromList</a></strong>: <em class="summary">This Program takes a list of IDs and extract the columns with corresponding IDs from a table</em>    </li>
    <li> <strong class="uidlink"><a href="extractFromProteineOrtho-module.html">extractFromProteineOrtho</a></strong>: <em class="summary">This Programme take proteineOrtho output and take file with Orthologue 1/1</em>    </li>
    <li> <strong class="uidlink"><a href="extractSeqFasta-module.html">extractSeqFasta</a></strong>: <em class="summary">This Programme extract Fasta Seq with liste keep</em>    </li>
    <li> <strong class="uidlink"><a href="extractSeqFastaCorresponding-module.html">extractSeqFastaCorresponding</a></strong>: <em class="summary">This Programme extract Fasta Seq with liste keep</em>    </li>
    <li> <strong class="uidlink"><a href="extractSeqFastaFromLen-module.html">extractSeqFastaFromLen</a></strong>: <em class="summary">This Programme filter sequences by length</em>    </li>
    <li> <strong class="uidlink"><a href="filterTabToSNP-module.html">filterTabToSNP</a></strong>: <em class="summary">SNPwithMissing: only line with one or more SNP with N</em>    </li>
    <li> <strong class="uidlink"><a href="getdistrgenotypeTAB-module.html">getdistrgenotypeTAB</a></strong>: <em class="summary">This Programme take proteineOrtho output and take file with Orthologue 1/1</em>    </li>
    <li> <strong class="uidlink"><a href="grepMotifFromAlignment-module.html">grepMotifFromAlignment</a></strong>: <em class="summary">This Programme parse Aligment info to build motif table of SNP in gene's</em>    </li>
    <li> <strong class="uidlink"><a href="grepUTRgene-module.html">grepUTRgene</a></strong>: <em class="summary">This Programme build fasta of listed genes with UTR size</em>    </li>
    <li> <strong class="uidlink"><a href="make_hapmapfile2-module.html">make_hapmapfile2</a></strong>: <em class="summary">This program takes a SNP matrix with no missing data and returns a hapmap file.</em>    </li>
    <li> <strong class="uidlink"><a href="make_ldhatfiles-module.html">make_ldhatfiles</a></strong>: <em class="summary">This Program takes a tab file and returns LDhat .sites and .locs files</em>    </li>
    <li> <strong class="uidlink"><a href="make_structure_dir-module.html">make_structure_dir</a></strong>: <em class="summary">This Programme make arborescence of rep of programme structure</em>    </li>
    <li> <strong class="uidlink"><a href="mappingOnRef-module.html">mappingOnRef</a></strong>: <em class="summary">This Programme map fasta with bwa mem on reference file</em>    </li>
    <li> <strong class="uidlink"><a href="mergeTabFiles-module.html">mergeTabFiles</a></strong>: <em class="summary">This Programme merged files (compress or not) into one file (compress or not)</em>    </li>
    <li> <strong class="uidlink"><a href="parseWithEgglib-module.html">parseWithEgglib</a></strong>: <em class="summary">This Program takes fasta alignement and check SNP info</em>    </li>
    <li> <strong class="uidlink"><a href="plot_barplot_DAPC-module.html">plot_barplot_DAPC</a></strong>: <em class="summary">This Programme Plot DAPC comboplot of csv file</em>    </li>
    <li> <strong class="uidlink"><a href="renameFile-module.html">renameFile</a></strong>: <em class="summary">This Programme rename files into a directory</em>    </li>
    <li> <strong class="uidlink"><a href="run_Assembly-module.html">run_Assembly</a></strong>: <em class="summary">This Programme run assembly of Jerome Gouzi pipeline for fastq file</em>    </li>
    <li> <strong class="uidlink"><a href="run_multiblast-module.html">run_multiblast</a></strong>: <em class="summary">This Programme run 1 blast for file include in path.</em>    </li>
    <li> <strong class="uidlink"><a href="run_pyinstaller-module.html">run_pyinstaller</a></strong>: <em class="summary">This run pyinstaller to build file on OS</em>    </li>
    <li> <strong class="uidlink"><a href="run_raxml-module.html">run_raxml</a></strong>: <em class="summary">This Programme run job array to lunch raxml</em>    </li>
    <li> <strong class="uidlink"><a href="splitMultiFasta-module.html">splitMultiFasta</a></strong>: <em class="summary">This Programme split multi-fasta to one file/sequences</em>    </li>
    <li> <strong class="uidlink"><a href="struc2runClumpak-module.html">struc2runClumpak</a></strong>: <em class="summary">This Programme take directory with output of structure repository and make file to run Clumpak</em>    </li>
    <li> <strong class="uidlink"><a href="tabToFasta-module.html">tabToFasta</a></strong>: <em class="summary">This Programme build fasta from tab file Adapt the RAM according to the size of the array (eg for 33127958 line and 48 individuals uses 34GB of RAM)</em>    </li>
    <li> <strong class="uidlink"><a href="vcfSummaryDepthandRatioAD-module.html">vcfSummaryDepthandRatioAD</a></strong>: <em class="summary">Programme plot histogram of depth and ratio nbReadsRef/nbReadsAlt into one pdf file</em>    </li>
    <li> <strong class="uidlink"><a href="vcfToTab-module.html">vcfToTab</a></strong>: <em class="summary">This Programme read vcf file (compress or not) and build filter SNP tab</em>    </li>
    <li> <strong class="uidlink"><a href="vcfToTabSplit-module.html">vcfToTabSplit</a></strong>: <em class="summary">This Programme build SNP tab with only python</em>    </li>
</ul>
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">

  <!-- Tree link -->
      <th bgcolor="#70b0f0" class="navbar-select"
          >&nbsp;&nbsp;&nbsp;Trees&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

      <th class="navbar" width="100%"></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Thu Jun  8 13:46:49 2017
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
